### DISCUSSION TITLE
ports/nrf: bluetooth module implementation

### DISCUSSION THREAD

**Discussion by yasnim24** [micropython/micropython#15453] [nRF]
Hello,

i've implemented a bluetooth module for the nRF port.

As I have now discovered there is already a pull request #8318
"nrf/bluetooth: Add support for nimble based ubluetooth."

I don't want to compete with that but if the community is interested I could pull that implementation.

**Background**

I already has developed an application based on aioble and therefore the bluetooth module. This application was running on the Unix port successfully. Now I wanted to run that application on the "Arduino nano 33 BLE" board.
Soon I realized that the nRF port has an ubluepy but no bluetooth module.

Therefore I decided to implement an bluetooth module which I could use for the "Arduino nano 33 BLE" board.

**Implemented features**

The following features are implemented and rudimentary tested with the "Arduino nano 33 BLE" board:
- Observer Role (Scanner)
- Central Role
- GATT Client
- Pairing and bonding in the Central Role; based on several components from the nRF SDK

The following features are potentially implemented but not tested:
- Broadcaster Role (Advertiser)
- Peripheral Role
- GATT Server
- Pairing and bonding in the Peripheral Role; based on several components from the nRF SDK

**Comment by jonnor** [comment]
Hi @yasnim24. I believe that @andrewleech, who made that MR, has said he is not using NRF so much these days (ref https://github.com/micropython/micropython/issues/16088#issuecomment-2439795059). So if you have the opportunity to contribute to updating that MR (probably create a new branch/MR) and drive it to completion, that would be fantastic.

**Reply by yasnim24** [reply]
Hello @jonnor, unfortunately I see no opportunity at the moment.

**Comment by jonnor** [comment]
@yasnim24 also, is your "bluetooth" module code available anywhere?

**Reply by yasnim24** [reply]
Hello @jonnor,
the code is now available here:
https://github.com/yasnim24/micropython/tree/ports_nrf_bluetooth_sdk_17_1_0
I didn't made it public so far because it lacks documentation etc.
